src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -2025,146 +1982,188 @@ else if (compareAndSetCtl(c, c) && casRunState(e, e | STOP)) {
-                            q.updateBase(nb);
-                            next = RESCAN | ((window << 16) & HMASK) | j;
-                            if (window != next && a[nk] != null)
-                                signalWork(a, nk); // limit propagation
-                            if (w != null)        // always true
-                                w.topLevelExec(t, q, j);
-                            break outer;
-                        }
-                        if (o == null) {
-                            if (next >= 0L && a[nk] != null)
-                                next |= RESCAN;
-                            break;
+    private int deactivate(WorkQueue w, int r, int phase) {
+        int p = phase | IDLE, active = phase + (IDLE << 1);
+        if (w != null) {                         // always true
+            w.phase = p;
+            long pc = ctl, qc;
+            for (;;) {                           // try to enqueue
+                w.stackPred = (int)pc;           // set ctl stack link
+                qc = (active & LMASK) | ((pc - RC_UNIT) & UMASK);
+                if (pc == (pc = compareAndExchangeCtl(pc, qc)))
+                    break;
+                if ((pc & RC_MASK) >= (qc & RC_MASK)) {
+                    p = w.phase = phase;         // back out on possible signal
+                    break;
+                }
+            }
+            if (p != phase &&                    // check quiescent termination
+                ((runState & SHUTDOWN) == 0 || quiescent() <= 0)) {
+                WorkQueue[] qs;
+                int spins = ((short)(qc >>> TC_SHIFT) << 1) + SPIN_WAITS + 1;
+                while (((p = w.phase) & IDLE) != 0 && --spins > 0)
+                    Thread.onSpinWait();         // reduce flailing
+                if ((p & IDLE) != 0 && (qs = queues) != null) {
+                    int n = qs.length, step = (r >>> 16) | 1;
+                    for (int i = r, l = n; l > 0; --l, i += step) {
+                        WorkQueue q;             // check for missed signals
+                        if ((q = qs[i & (n - 1)]) != null &&
+                            q.top - q.base > 0) {
+                            if (ctl == qc && compareAndSetCtl(qc, pc)) {
+                                p = w.phase = active;
+                                break;           // self-signal
+                            }
+                            if (((p = w.phase) & IDLE) == 0)
+                                break;
                         }
                     }
+                    if ((p & IDLE) != 0) {
+                        long delay = (((qc & RC_MASK) > 0L) ? 0L :
+                                      (w.source != INVALID_ID) ? keepAlive :
+                                      TIMEOUT_SLOP); // minimal delay if cascade
+                        if (((p = w.phase) & IDLE) != 0)
+                            p = awaitWork(w, delay); // block, drop, or exit
+                    }
```

Comment:
I'm presuming the code below would be equivalent but avoid calculating the delay unless w.phase is IDLE?

```suggestion
                    if ((p & IDLE) != 0 && ((p = w.phase) & IDLE) != 0) {
                        long delay = (((qc & RC_MASK) > 0L) ? 0L :
                                      (w.source != INVALID_ID) ? keepAlive :
                                      TIMEOUT_SLOP); // minimal delay if cascade
                            p = awaitWork(w, delay); // block, drop, or exit
                    }
```

Reply:
Seems slightly better not to recheck until call, since previous loop would usually have caught this if there was actually any work.